### PR TITLE
Add identity expansion checking in ill-form expansion 

### LINKED ISSUE #3642: Macro expansion loops
This code give 100% CPU usage:

image

cc @flodiebold, @edwin0cheng, this is quite likely related to #3593

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR try to add more checking code in error case in macro expansion. The bug in #3642 is introduced by  #3580 , which allow ill-form macro expansion in *all* kind of macro expansions. 

In general we should separate hypothetical macro expansion and the actual macro expansion call. However, currently the `Semantic`  workflow we are using only support single macro expansion type, we might want to review it and make it works in both ways. (Maybe add a field in `MacroCallLoc` for differentiation) 

Fix #3642

### CODE REVIEW

**Comment by edwin0cheng** [comment]
@flodiebold

**Comment by bjorn3** [comment]
I think you meant to use #3642, instead of `@3642`.

**Comment by edwin0cheng** [comment]
@bjorn3 oops.. not enough coffee, thanks <3

**Comment by bjorn3** [comment]
:coffee:

File: crates/ra_ide/src/diagnostics.rs [hunk 1]
```diff
@@ -715,4 +715,34 @@ fn main() {
             check_struct_shorthand_initialization,
         );
     }
+
+    #[test]
+    fn test_bad_macro_stackover() {
```
Comment (flodiebold) [inline crates/ra_ide/src/diagnostics.rs]: could this test be in `hir_ty` instead?

**Comment by flodiebold** [comment]
Thanks for looking into this by the way!

File: crates/ra_ide/src/diagnostics.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/ra_ide/src/diagnostics.rs]: Sure. 

Originally I found out the bug is triggered in diagnostics stage so I put the test in there but hir_ty is a better place to put.

**Comment by edwin0cheng** [comment]
> Hmm looks good to me, we should also make sure that we're limiting the expansion depth everywhere though, right? I.e. it seems to me we're probably not doing that in expression macro expansion?

Yea, good idea !! I will do another PR for that !

**Comment by edwin0cheng** [comment]
bors retry

**Pull request merged**